client/src/main/java/org/apache/rocketmq/client/impl/producer/DefaultMQProducerImpl.java

```diff
@@ -695,16 +695,13 @@ private SendResult sendDefaultImpl(
             if (sendResult != null) {
                 return sendResult;
             }
+            StringBuilder info = new StringBuilder(300);
+            info.append("Send [").append(times).append("] times, still failed, cost [")
+                .append(System.currentTimeMillis() - beginTimestampFirst).append("]ms, Topic: ").append(msg.getTopic())
+                .append(", BrokersSent: ").append(Arrays.toString(brokersSent))
+                .append(FAQUrl.suggestTodo(FAQUrl.SEND_MSG_FAILED));
```

Comment:
> It seems that the String.format is more readable here

This optimization has been done from a performance perspective. The performance of String.format is relatively poor and triggers an array expansion, which can be referenced in issue #5804 . However, readability has been reduced.Performance enhancement should be considered as the first priority.

Reply:
IMO, these transitions are all in the exception scene,

it will lost a little performance when in the little scene,

so  we can choose the readability when in this scene.

Reply:
Use MoreObjects.toStringHelper of guava or ToStringBuilder of commons-lang is more readable